### PR TITLE
[MIG] migrate product_creation_prevention from v12 to v15

### PR DESCRIPTION
**Pull request by dsolanki-initos**

### CODE REVIEW

**Comment by jans23** [comment]
`.pre-commit-config.yaml` shouldn't be part of this PR.

**Comment by dsolanki-initos** [comment]
@jans23  pre-commit was failing so I have rebased this branch with 15.0 which contains a new fix of pre-commit-config.yaml

**Comment by jans23** [comment]
Has the rebase been executed properly? Because the changes shown in this PR are in 15.0 branch already.

**Comment by dsolanki-initos** [comment]
> Has the rebase been executed properly? Because the changes shown in this PR are in the 15.0 branch already.

Yes rebase I executed it correctly, but after the rebase with 15.0 all the recent changes you have merged with 15.0 will show in this branch.

**Review by manengel — CHANGES_REQUESTED** [review]
I installed the module on my local system and it seemingly has no effect. Also, where did the view modifications of Odoo 12 go to? (https://github.com/Nitrokey/odoo-modules/pull/97)

**Comment by dsolanki-initos** [comment]
> I installed the module on my local system and it seemingly has no effect. Also, where did the view modifications of Odoo 12 go to? (#97)

Did you check with the other internal or admin user? Because this module prevents the creation of a new record of products from the other internal users, not the admin user.

**Comment by manengel** [comment]
> > I installed the module on my local system and it seemingly has no effect. Also, where did the view modifications of Odoo 12 go to? (#97)
> 
> Did you check with the other internal or admin user? Because this module prevents the creation of a new record of products from the other internal users, not the admin user.

I did both actually. But I will re-check if the other internal user possibly had too many permissions.

**Comment by dsolanki-initos** [comment]
> > > I installed the module on my local system and it seemingly has no effect. Also, where did the view modifications of Odoo 12 go to? (#97)
> > 
> > 
> > Did you check with the other internal or admin user? Because this module prevents the creation of a new record of products from the other internal users, not the admin user.
> 
> I did both actually. But I will re-check if the other internal user possibly had too many permissions.

I have attached a video for the same, Please go through it once


https://user-images.githubusercontent.com/50451593/205032033-ceb480c8-e43e-4c51-8f1b-5cdf8c0eb4f2.mp4

**Comment by manengel** [comment]
> > > > I installed the module on my local system and it seemingly has no effect. Also, where did the view modifications of Odoo 12 go to? (#97)
> > > 
> > > 
> > > Did you check with the other internal or admin user? Because this module prevents the creation of a new record of products from the other internal users, not the admin user.
> > 
> > 
> > I did both actually. But I will re-check if the other internal user possibly had too many permissions.
> 
> I have attached a video for the same, Please go through it once
> product_creation_prevention_vid.mp4

I tried to do that. Can you somehow make a list of what access rights were set for this user? Because, it seemingly does not matter how many I remove, my other user has seemingly always too many permissions...also tried to toggle the access rights, set base.group_user as group etc. and it did all not help. So I think there must have been a change from Odoo 12 to 15 which is essentially a breaking one. I will attempt this again with a fresh database and then will report if this was really just a local phenomenom. But I cannot imagine that because the access rights definitely were entered correctly in the database.

EDIT: I did now re-check with an empty database, a new normal user and the module being freshly installed and the access rights set as thought of. No change on my side.

**Comment by dsolanki-initos** [comment]
> > > > > I installed the module on my local system and it seemingly has no effect. Also, where did the view modifications of Odoo 12 go to? (#97)
> > > > 
> > > > 
> > > > Did you check with the other internal or admin user? Because this module prevents the creation of a new record of products from the other internal users, not the admin user.
> > > 
> > > 
> > > I did both actually. But I will re-check if the other internal user possibly had too many permissions.
> > 
> > 
> > I have attached a video for the same, Please go through it once
> > product_creation_prevention_vid.mp4
> 
> I tried to do that. Can you somehow make a list of what access rights were set for this user? Because, it seemingly does not matter how many I remove, my other user has seemingly always too many permissions...also tried to toggle the access rights, set base.group_user as group etc. and it did all not help. So I think there must have been a change from Odoo 12 to 15 which is essentially a breaking one. I will attempt this again with a fresh database and then will report if this was really just a local phenomenom. But I cannot imagine that because the access rights definitely were entered correctly in the database.
> 
> EDIT: I did now re-check with an empty database, a new normal user and the module being freshly installed and the access rights set as thought of. No change on my side.

You can refer the access options of the user from  the attached image

Odoo - Marc Demo

**Comment by manengel** [comment]
@dsolanki-initos I made a very, very interesting find in the context of our discussion above. For some reasons, the rights per app (tested with Sales as example, same as you did) can somehow surpass the limitation in rights made by our module, are in fact kinda in conflict with that. For the Sales workflow, it is seemingly enough to not give the user Administrator rights in the Sales category. But for Invoicing and Inventory, this seems more complicated. So, if I use EXACTLY the same rights as you did, then the module is doing its job within the Sales Workflow, but not anywhere else. Do you have an idea how we can possibly handle this situation?

**Comment by dsolanki-initos** [comment]
> @dsolanki-initos I made a very, very interesting find in the context of our discussion above. For some reasons, the rights per app (tested with Sales as example, same as you did) can somehow surpass the limitation in rights made by our module, are in fact kinda in conflict with that. For the Sales workflow, it is seemingly enough to not give the user Administrator rights in the Sales category. But for Invoicing and Inventory, this seems more complicated. So, if I use EXACTLY the same rights as you did, then the module is doing its job within the Sales Workflow, but not anywhere else. Do you have an idea how we can possibly handle this situation?


As per the above image user's access rights, user can not create a product anywhere, I have checked other workflow as well, For same, I have attached a video for reference, Please go through it.

https://user-images.githubusercontent.com/50451593/205281565-a6fcc94b-4474-4a3a-b676-9ec87754f73f.mp4

**Review by manengel — COMMENTED** [review]
I managed to make the module work locally. But I noticed one thing: It does only work under the condition that the user is not an Administrator of the apps Sales, Invoicing, etc. @jans23 Is it okay from your side that the module works in that way?

**Comment by jans23** [comment]
hmm, it would be better if this rule applies for Administrators too. How could this be made possible? If it's too complicated, we can postpone it as a future improvement.

**Comment by manengel** [comment]
> hmm, it would be better if this rule applies for Administrators too. How could this be made possible? If it's too complicated, we can postpone it as a future improvement.

This should be possible by overriding the rights given to the respective Administrator groups. 
So, one must look inside the code of the sales, invoicing... modules to find the respective XML IDs of the user groups and then put more restrictive rules than already present inside this module's ir.model.access.csv

**Comment by dsolanki-initos** [comment]
@manengel @jans23 Update the access right to prevent product record creation for the administrator group of sales, invoice, purchase, inventory, and manufacture. Please check it.

**Comment by jans23** [comment]
I think the module description needs to be updated too.

**Comment by manengel** [comment]
@dsolanki-initos (@jans23) I did test again. But for some reason, the new modifications have no effect.

**Comment by dsolanki-initos** [comment]
Hello @manengel, I've checked in the local and dob environment as well, I've attached a video of the same please refer it. 

https://user-images.githubusercontent.com/50451593/206094179-aadc2c19-7a96-47ac-ae82-60912f85ff32.mp4

File: product_creation_prevention/__manifest__.py [hunk 1]
```diff
@@ -1,9 +1,8 @@
 {
     "name": "Product Creation Prevention",
     "version": "15.0.1.0.0",
-    "summary": """Product Creation Prevention
-    By adding access right we are preventing users to create products 'on the fly'
-    """,
+    "summary": """Using this module we can prevent to create new products by
+    users except admin user""",
```
Comment (jans23) [inline product_creation_prevention/__manifest__.py:5]: I understood that product creation prevention now also applies to admin users. Doesn't it?

**Comment by manengel** [comment]
> Hello @manengel, I've checked in the local and dob environment as well, I've attached a video of the same please refer it.

I am trying that right now again, will let you know. It seems that for some reason, this module does technically not work as it should without me resetting the entire database locally.

**Review by manengel — APPROVED** [review]
@dsolanki-initos I do not quite understand what is causing exactly that module to cause so many issues locally. But I basically see it as proven that it does work.

**Pull request merged**